src/cuda_block_manager.cc

```diff
@@ -0,0 +1,149 @@
+
+Status
+CudaBlockManager::Create(double min_supported_compute_capability)
+{
+  if (instance_ != nullptr) {
+    return Status(
+        Status::Code::INTERNAL,
+        (std::string("CUDA Block manager has already been created.")));
+  }
+
+  std::set<int> supported_gpus;
+  RETURN_IF_ERROR(
+      GetSupportedGPUs(&supported_gpus, min_supported_compute_capability));
+
+  size_t block_size;
+  RETURN_IF_ERROR(GetAllocationGranularity(block_size));
+  instance_.reset(new CudaBlockManager());
+
+  instance_->block_size_ = block_size;
+  for (auto gpu : supported_gpus) {
+    instance_->free_blocks_[gpu] = std::vector<CUmemGenericAllocationHandle>();
+  }
+
+  return Status::Success;
+}
+
+Status
+CudaBlockManager::Allocate(
+    size_t size, std::unique_ptr<Allocation>& allocation, int device_id)
+{
+  if (instance_ == nullptr) {
+    return Status(
+        Status::Code::INTERNAL,
+        (std::string("CUDA Block manager has not been created.")));
+  }
+
+  std::lock_guard<std::mutex> lock(instance_->mu_);
+  if (instance_->free_blocks_.find(device_id) ==
+      instance_->free_blocks_.end()) {
+    return Status(
+        Status::Code::INTERNAL,
+        (std::string("Invalid device id '") + std::to_string(device_id) +
+         "' to allocate memory from."));
+  }
+
+  // Align the byte size to number of blocks.
+  size_t num_blocks =
+      (size + instance_->block_size_ - 1) / (instance_->block_size_);
+  while (num_blocks > 0) {
+    if (instance_->free_blocks_[device_id].size() > 0) {
+      allocation->AddBlock(instance_->free_blocks_[device_id].back());
+      instance_->free_blocks_[device_id].pop_back();
+    } else {
+      CUmemGenericAllocationHandle block{};
+      CUmemAllocationProp prop = {};
+      prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+      prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+      prop.location.id = device_id;
+      RETURN_IF_CUDA_DRIVER_ERR(
+          cuMemCreate(&block, instance_->block_size_, &prop, 0 /* flags */),
+          std::string("cuMemCreate failed:"));
+      allocation->AddBlock(block);
+    }
+    num_blocks--;
+  }
+
+
+  return Status::Success;
+}
+
+Status
+CudaBlockManager::Free(
+    std::vector<CUmemGenericAllocationHandle>&& blocks, int device_id)
+{
+  if (instance_ == nullptr) {
+    return Status(
+        Status::Code::INTERNAL,
+        (std::string("CUDA Block manager has not been created.")));
+  }
+
+  std::lock_guard<std::mutex> lock(instance_->mu_);
+
+  if (instance_->free_blocks_.find(device_id) ==
+      instance_->free_blocks_.end()) {
+    return Status(
+        Status::Code::INTERNAL,
+        (std::string("Invalid device id '") + std::to_string(device_id) +
+         "' to allocate memory from."));
+  }
+
+  auto& free_blocks = instance_->free_blocks_[device_id];
+  free_blocks.insert(free_blocks.end(), blocks.begin(), blocks.end());
+
+  return Status::Success;
+}
+
+CudaBlockManager::~CudaBlockManager()
+{
+  for (auto& [device, free_blocks] : free_blocks_) {
```

Comment:
Isn't this C++17 structured bindings? I guess we're implicitly compiling with 17 on ubuntu 22.04.

Reply:
I'm curious if this part will compile on Windows. I think the rest of the VMM API stuff used here is cross-platform without the IPC components.

Reply:
Good point fixed.